### PR TITLE
8312019: Simplify and modernize java.util.BitSet.equals

### PR DESCRIPTION
**Pull request by pavelrappo**
Please review this PR to use modern APIs and language features to simplify equals for BitSet. 

I couldn't see how to refactor hashCode using Arrays utility methods in a way that preserves its specification. So, aside from refactoring its doc comment and annotating it with `@Overrides`, I left it intact.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312019](https://bugs.openjdk.org/browse/JDK-8312019): Simplify and modernize java.util.BitSet.equals (**Bug** - P4)

### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to [39de7d2a](https://git.openjdk.org/jdk/pull/14868/files/39de7d2ae6323c7e1b9dc1f38785fc8f79438e05)
 * [Martin Buchholz](https://openjdk.org/census#martin) (@Martin-Buchholz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14868/head:pull/14868` \
`$ git checkout pull/14868`

Update a local copy of the PR: \
`$ git checkout pull/14868` \
`$ git pull https://git.openjdk.org/jdk.git pull/14868/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14868`

View PR using the GUI difftool: \
`$ git pr show -t 14868`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14868.diff">https://git.openjdk.org/jdk/pull/14868.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14868#issuecomment-1634037680)

### CODE REVIEW

**Review by RogerRiggs — APPROVED** [review]
LGTM

File: src/java.base/share/classes/java/util/BitSet.java [hunk 1]
```diff
@@ -1015,8 +1015,10 @@ public void andNot(BitSet set) {
     }
 
     /**
-     * Returns the hash code value for this bit set. The hash code depends
-     * only on which bits are set within this {@code BitSet}.
+     * {@return the hash code value for this bit set}
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/BitSet.java:1018]: I tried to find the (non-obvious) spec for {@return ...}
Googling led me to the now embarrassingly outdated
https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html
https://www.oracle.com/java/technologies/javase/javadoc-tool.html#javadocdocuments
before finding the "real" well-maintained doc
https://docs.oracle.com/en/java/javase/20/docs/specs/javadoc/doc-comment-spec.html
(and that would benefit from TOC and inline anchors)

Improving these docs might be the most valuable engineering work that javadoc team could do!

**Comment by pavelrappo** [comment]
> BitSet#equals

Did you mean BitSet#hashCode?

**Comment by Martin-Buchholz** [comment]
> > BitSet#equals
> 
> Did you mean BitSet#hashCode?

No. BitSet#equals uses the private fields and methods of its argument, which OO purists would never allow.
If the other is a subclass with a different private representation, this code would fail.
(Don't fix it!)

**Comment by pavelrappo** [comment]
> > > BitSet#equals
> > 
> > 
> > Did you mean BitSet#hashCode?
> 
> No. BitSet#equals uses the private fields and methods of its argument, which OO purists would never allow. If the other is a subclass with a different private representation, this code would fail. (Don't fix it!)

I see.

While both `hashCode` and `equals` use private fields, `equals` also calls the `checkInvariants` private method on `this` and the other set. I reckon, it makes `equals` quantitatively worse than `hashCode`, rather than qualitatively.

**Comment by pavelrappo** [comment]
@RogerRiggs, @Martin-Buchholz, are you okay with a few redundant range checks?

`Arrays.equals(a, aFromIndex, aToIndex, b, bFromIndex, bToIndex)` performs two range checks, which aren't needed here. However, `Arrays.equals` is readable; a less readable but leaner alternative would be `ArraysSupport.mismatch(a, aFromIndex, b, bFromIndex, length)`. Here's a diff from this PR's version, to consider:

```
diff --git a/src/java.base/share/classes/java/util/BitSet.java b/src/java.base/share/classes/java/util/BitSet.java
index 7189dc49228..fea9693d7ed 100644
--- a/src/java.base/share/classes/java/util/BitSet.java
+++ b/src/java.base/share/classes/java/util/BitSet.java
@@ -33,6 +33,8 @@ import java.util.function.IntConsumer;
 import java.util.stream.IntStream;
 import java.util.stream.StreamSupport;
 
+import jdk.internal.util.ArraysSupport;
+
 /**
  * This class implements a vector of bits that grows as needed. Each
  * component of the bit set has a {@code boolean} value. The
@@ -1076,8 +1078,11 @@ public class BitSet implements Cloneable, java.io.Serializable {
         checkInvariants();
         set.checkInvariants();
 
+        if (wordsInUse != set.wordsInUse)
+            return false;
+
         // Check words in use by both BitSets
-        return Arrays.equals(words, 0, wordsInUse, set.words, 0, set.wordsInUse);
+        return ArraysSupport.mismatch(words, 0, set.words, 0, wordsInUse) == -1;
     }
 
     /**
```

**Comment by Martin-Buchholz** [comment]
TIL about toLongArray.. BitSet#equals would be more correct if it called set.toLongArray, but at a cost of an extra copy.

---

I appreciate the desire to call checkInvariants; I've written some checkInvariants methods myself, but I comment  out any calls to them in the production copy of the code.  Optional runtime checking has not really succeeded in Java.

**Comment by Martin-Buchholz** [comment]
Performance of core classes is very important, - I am willing to have less elegant code with better performance.
Most users would also prefer faster code "for free".
I would prefer to comment out the checkInvariants() and to not introduce new checks.
I don't trust the VM to optimize away all the trivially true checks.

But I've seen other maintainers opt for elegant cleaner slower code instead, and I've tried not to complain!

**Comment by RogerRiggs** [comment]
Using ArraySupport.mismatch is pretty straightforward and avoids the duplicate checks.  That fine by me.

**Review by Martin-Buchholz — APPROVED** [review]
LGTM

**Pull request closed without merging**